### PR TITLE
Reimplement prometheus receiver

### PR DESCRIPTION
**Pull request by fivesheep**
## Problems with the current implementation of prometheus receiver
The current prometheus receiver is based on the https://github.com/orijtech/promreceiver  library. in the past couple of months, we have found a couple of issues with this implementation and has reported either in gitter or the library's git repo, which including: 

- dose not support all the data types prometheus: https://github.com/orijtech/promreceiver/issues/5
- throwing error when encounter unknown metrics type which caused the rest of the metric points from the prometheus scrape query to be lost: https://github.com/orijtech/promreceiver/issues/3
- metric labels not handle properly:  https://github.com/orijtech/promreceiver/issues/1
- did not handle NaN values properly: https://github.com/orijtech/promreceiver/issues/2

## Why reimplementing instead of submitting fixes
So far, some of the above issues were fixed, some not. However, this is not the whole story. After diving deeper into the code, we have found that the the library actually has a number of serious bugs, we conclude that without a complete rewrite, it's not going to work properly:

1. A global appender is used, which can easily run into race conditions when multiple scrape target is configured
2. Failed to group metrics in the original form, such as a single group of metric from prometheus can be divided into multiple. This is a big issue if one wants to config a prometheus exporter on the other end of the agent to a prometheus server, as metric group with same name is not allowed.
3. The library is trying to cache all the data scraped from remote metrics endpoints, then feed the delta to downstream, this doesn't seem like what is expected. even worst, it has used a global cache to store the data, which can be easily corrupted when scraping multiple targets
4. The library is also trying to use the cache to generate STDDEV for histogram which is a mission impossible as the original data doesn't provide any original sampled points. 

More details for this PR trying to address can be found in the following document, which is also part of the PR: https://github.com/fivesheep/opencensus-service/blob/reimplement-prometheus-receiver/receiver/prometheusreceiver/README.md

## an example of issues (NaN) values:

with the following ocagent config
```yaml
receivers:
    prometheus:
        config:
            global:
                scrape_interval: 120s
                scrape_timeout: 8s
            scrape_configs:
            - job_name: 'cadvisor'
              scrape_interval: 15s
              static_configs:
              - targets: ['localhost:8888']

exporters:
    prometheus:
            namespace: "promdemo"
            address: "localhost:8889"
```

compare the outputs from original prometheus endpoint and the ocagent prometheus exporter

original (cadvisor)
```
...
# HELP process_resident_memory_bytes Resident memory size in bytes.
# TYPE process_resident_memory_bytes gauge
process_resident_memory_bytes 5.6446976e+07
# HELP process_start_time_seconds Start time of the process since unix epoch in seconds.
# TYPE process_start_time_seconds gauge
process_start_time_seconds 1.56037276274e+09
# HELP process_virtual_memory_bytes Virtual memory size in bytes.
# TYPE process_virtual_memory_bytes gauge
process_virtual_memory_bytes 1.501503488e+09
```

from ocagent promethues exporter
```
...
# HELP promdemo_process_resident_memory Resident memory size in bytes.
# TYPE promdemo_process_resident_memory counter
promdemo_process_resident_memory NaN
# HELP promdemo_process_start_time Start time of the process since unix epoch in seconds.
# TYPE promdemo_process_start_time counter
promdemo_process_start_time NaN
# HELP promdemo_process_virtual_memory Virtual memory size in bytes.
# TYPE promdemo_process_virtual_memory counter
promdemo_process_virtual_memory NaN
```

### CODE REVIEW

**Comment by pjanotti** [comment]
Hi @fivesheep thanks for providing the context for this PR. It brings the whole functionality so it is on the larger side. We will need a bit of time to review all pieces, but, I will try to provide some of the feedback as we go.

**Comment by songy23** [comment]
Agree this PR is a bit too large. Consider splitting it to multiple smaller PRs, like one for `metricbuilder`, one for logger, etc.

**Comment by odeke-em** [comment]
Thank you for working on this @fivesheep and for the fixes and overhaul! @pjanotti pinged me about it and my apologies for not being responsive on my repo as I've been swamped with work. I think that to get this implementation in here, some few suggestions:
a) @fivesheep perhaps you can put the new exporter under your repos, license it for OpenCensus Authors under Apache 2.0 for the time being and when @songy23 has finished creating a repository under https://github.com/census-ecosystem it can be migrated there and reviewed independently
b) Concurrently with step `a)` @songy23 please help create the vanity URL `contrib.go.opencensus.io/receiver/prometheus` and look at adding @fivesheep's exporter behind that 
c) When https://github.com/census-ecosystem/opencensus-go-prometheus-receiver has been created, then you can start transferring the code piecemeal and have the respective small code reviews
d) When step `c)` is completed, the new receiver can reside behind the vanity URL and the repository from `a)` can be sunset.

I say this because the logic will take some time to review but during that this opencensus-service will have lots of changes in it. Working on it independently ensures you won't be blocked, you can iterate faster, make test suites etc but also that correctness supersedes :)

Hope that this can help with increasing the velocity of getting things in.

**Comment by fivesheep** [comment]
@pjanotti @songy23 and @odeke-em  thanks for the response. I would love to create an individual repo to host the library, however, I cannot make this decision by myself because of my company's opensource policy.  I would need to ask our opensource committee if I am allowed to do that, or I need to go through the other opensource request, which can take a very long time to get approval.

**Comment by pjanotti** [comment]
@fivesheep we would like to eventually make this part of the core as we are going to leverage OC service to be the starting point of https://github.com/open-telemetry/opentelemetry-service/tree/master/docs and Prometheus is part of the ciore. The contrib was proposed as a temporary stage to make things speedier meanwhile, if that is not the case then we make this work directly on core.

File: receiver/prometheusreceiver/README.md [hunk 1]
```diff
@@ -0,0 +1,452 @@
...
+        Points: []*metricspb.Point{
+          {Timestamp: tsOc, Value: &metricspb.Point_DoubleValue{DoubleValue: 3.0}},
+        },
+      },
+    },
+  },
+}
+```
+
+*Note: `tsOc` is a timestamp object representing the current ts*
```
[the remark this answers is not in the document]
Comment (fivesheep) [reply receiver/prometheusreceiver/README.md]: You are right, the `current ts` here was actually referring to the same thing, I will update the description as per your comment to reduce the confusion.

I was trying to differentiate timestamps provided from the original metric endpoint and the ones generated by the scraperLoop which you were referring to, and I meant `current`. 

An example of such metric output is shown below:
```
# HELP http_requests_total The total number of HTTP requests.
# TYPE http_requests_total counter
http_requests_total{method="post",code="200"} 1027 1395066363000
http_requests_total{method="post",code="400"}    3 1395066363000
```

once  the `honor_timestamps` flag of scrapping config is set to true, which is the default value, this timestamp is respected, the logic can be found from: https://github.com/prometheus/prometheus/blob/0c0638b080cf1565fc4e5b7ee3fd35c36ae5832a/scrape/scrape.go#L1084-L1091

In the implantation of prometheus receiver, whichever timestamp is provided from the Add/AddFast method will be used.

File: receiver/prometheusreceiver/README.md [hunk 2]
```diff
@@ -0,0 +1,452 @@
...
+    Timeseries: []*metricspb.TimeSeries{
+      {
+        StartTimestamp: tsOc,
+        LabelValues:    []*metricspb.LabelValue{{Value: "1", HasValue: true}, {Value: "bar", HasValue: true}},
+        Points: []*metricspb.Point{
+          {Timestamp: tsOc, Value: &metricspb.Point_DoubleValue{DoubleValue: 1.0}},
+        },
+      },
+      {
+        StartTimestamp: tsOc,
```
[the remark this answers is not in the document]
Comment (fivesheep) [reply receiver/prometheusreceiver/README.md]: Dose this logic also apply to Counters?

File: receiver/prometheusreceiver/internal/metricsbuilder.go [hunk 3]
```diff
@@ -0,0 +1,540 @@
...
+		if strings.HasSuffix(metricName, metricsSuffixSum) {
+			mg.sum = v
+			mg.hasSum = true
+		} else if strings.HasSuffix(metricName, metricsSuffixCount) {
+			mg.count = v
+			mg.hasCount = true
+		} else {
+			pg, ok := b.currentDpGroups[groupKey]
+			if !ok {
+				pg = make([]*dataPoint, 0, 4)
```
[the remark this answers is not in the document]
Comment (fivesheep) [reply receiver/prometheusreceiver/internal/metricsbuilder.go]: It's just some number in my mind trying to preallocate a slice, there's no reason to have this value. probably better just change it to 0 to avoid confusions

File: receiver/prometheusreceiver/metrics_receiver_test.go [hunk 4]
```diff
@@ -197,87 +196,188 @@ buffer_count: 2
...
 	}
 
-	wantMetricPb2 := &metricspb.Metric{
-		MetricDescriptor: &metricspb.MetricDescriptor{
-			Name:        "e2ereceiver_e2e_call_latency",
-			Description: "The latency in milliseconds per call",
-			Type:        metricspb.MetricDescriptor_CUMULATIVE_DISTRIBUTION,
-			LabelKeys: []*metricspb.LabelKey{
-				{Key: "method"},
+	want2 := []data.MetricsData{
```
Comment (rghetia) [inline receiver/prometheusreceiver/metrics_receiver_test.go]: want2 and want1 are same.

File: receiver/prometheusreceiver/metrics_receiver_test.go [hunk 5]
```diff
@@ -286,22 +386,61 @@ buffer_count: 2
...
+	// Since these tests rely on underdeterministic behavior and timing that's imprecise.
+	// The best that we can do is provide any of variants of what we want.
+	wantPermutations := [][]data.MetricsData{
+		want1, want2,
+	}
+
+	for _, want := range wantPermutations {
+		if !reflect.DeepEqual(got, want) {
+			t.Errorf("different metric got:\n%v\nwant:\n%v\n", string(exportertest.ToJSON(got)),
+				string(exportertest.ToJSON(want)))
```
Comment (rghetia) [inline receiver/prometheusreceiver/metrics_receiver_test.go]: is the purpose here to fail if either want1 or want2 doesn't match? (assuming that want1 and want2 are different. see prev comment. they don't seem to be different)

Comment (fivesheep) [reply receiver/prometheusreceiver/metrics_receiver_test.go]: This test was inherited from the original implementation of pmreceiver,  @odeke-em  might have more insight on this test.

File: receiver/prometheusreceiver/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (fivesheep) [reply receiver/prometheusreceiver/README.md]: Based on this comment, I have a feeling that are we expecting the counter to be a delta from OpenCensus in between two scrapes on the same target? If that's the case, the original implementation on Counter was correct, it did compute a delta with the previous value, however, it also computes delta on Gagues.  what about other data types, like histograms, the buckets also need to be a delta from the previous scrape?   @rghetia and @odeke-em  could you provide some specs on opencensus proto when a delta is needed, and when not, given that all sum/count/bucket_count values from prometheus are all cumulative.

[the remark this answers is not in the document]
Comment (rghetia) [reply receiver/prometheusreceiver/README.md]: For counters it is delta between first (not previous) and current. Same for histogram sum/count/bucket_count. For Gauges it should be reported as is.

[the remark this answers is not in the document]
Comment (fivesheep) [reply receiver/prometheusreceiver/README.md]: Thank you for the reply. I am still a bit confused on the `delta`, since in Prometheus the values are recorded since the application started, the first value will be 0, observed or not.  I have also created a demo app with oclib and observed what was fed to the exporters, code as shown below
```
func main() {
	key, _ := tag.NewKey("foo")
	counter := stats.Int64("my-counter", "counter", "1")
	countView := &view.View{
		Measure:     counter,
		Name:        "simple-counter",
		Description: "Counter",
		Aggregation: view.Count(),
		TagKeys:     []tag.Key{key},
	}

	view.Register(countView)
	exp, _ := ocagent.NewExporter(ocagent.WithInsecure(), ocagent.WithServiceName("opencensus-demo"))
	defer exp.Stop()
	view.RegisterExporter(exp)

	// For demo purposes we'll use a very short view data reporting
	// period so that we can examine the stats send to our exporter, quickly.
	view.SetReportingPeriod(15 * time.Second)

	ctx, _ := tag.New(context.Background(), tag.Upsert(key, "bar"))
	for i := int64(0); ; i++ {
		stats.Record(ctx, counter.M(1))
		<-time.After(500 * time.Millisecond)
	}
}
```

from the ocagent side, the following metrics were seen from 2 reporting cycles. The first one is:
```
{
        "metric_descriptor": {
          "name": "simple-counter",
          "description": "Counter",
          "unit": "1",
          "type": 4,
          "label_keys": [
            {
              "key": "foo"
            }
          ]
        },
        "timeseries": [
          {
            "start_timestamp": {
              "seconds": 1560978301,
              "nanos": 352044000
            },
            "label_values": [
              {
                "value": "bar",
                "has_value": true
              }
            ],
            "points": [
              {
                "timestamp": {
                  "seconds": 1560978301,
                  "nanos": 352102000
                },
                "Value": {
                  "Int64Value": 30
                }
              }
            ]
          }
        ]
      }
```

and in the 2nd one:

```
{
        "metric_descriptor": {
          "name": "simple-counter",
          "description": "Counter",
          "unit": "1",
          "type": 4,
          "label_keys": [
            {
              "key": "foo"
            }
          ]
        },
        "timeseries": [
          {
            "start_timestamp": {
              "seconds": 1560978301,
              "nanos": 352044000
            },
            "label_values": [
              {
                "value": "bar",
                "has_value": true
              }
            ],
            "points": [
              {
                "timestamp": {
                  "seconds": 1560978316,
                  "nanos": 351525000
                },
                "Value": {
                  "Int64Value": 60
                }
              }
            ]
          }
        ]
      }
```

The first one has value 30, the 2nd one has 60. Basically, it's the value since `start_time`. Same observation found on histrograms too. This means the implementation from this PR is closed to the behaviors of OCLib (go). however, the start_time need to be adjust, it has to be the timestamp when the metrics is first recorded on the original system, however, this timestamp is not provided, the only available value is the first given timestamp when a prometheus target is scraped, and it needs to be cached for all following scrapes. Since targets are separated application, they can go down and restart at any given time. the prometheus receiver also need a way to detect such events and resetting the start_time stamp. we can compare any monotonically increasing metrics from a target, any decreasing of these values can be used as a signal.   

@rghetia  do you think the above description will be the correct behavior?

[the remark this answers is not in the document]
Comment (rghetia) [reply receiver/prometheusreceiver/README.md]: What you have described is correct for the case 'a' below.

a. OcAgent is up and running. Demo app starts later and may or may not restart.
b. OcAgent restarts or starts after Demo App. 

In case 'b' there is no way to know when the Demo App started. May be it started few seconds ago or may be it started 10 days ago. If it started 10 days ago then it will show a big spike when rate/delta chart is plotted for the counter. For example, first value is 100000, second is 100010 after 1 minute, 100020 and so on. Delta chart would be flat line of 10 except for the first value which is 100000.

the other problem is that for backend like Stackdriver you cannot have timestamp of a point be the same as StartTime. Workaround there is to cache first timestamp after subtracting few seconds. 
Better solution to tackle both problem is to cache in the first value and timestamp. Do not report first value. Then report delta. You would still need to replace cached value and timestamp when  the value goes down (indicating that the demo app restarted).

[the remark this answers is not in the document]
Comment (fivesheep) [reply receiver/prometheusreceiver/README.md]: Nice, thanks for the clarification. would it be even better if we give an option to the user allow them to choose if they want to report the first value or not?

[the remark this answers is not in the document]
Comment (rghetia) [reply receiver/prometheusreceiver/README.md]: Probably not. Due to very large first values, graph is impossible to use. I have experienced that myself with different receiver.

File: receiver/prometheusreceiver/internal/metricsbuilder.go [hunk 6]
```diff
@@ -0,0 +1,549 @@
...
+	return lvs
+}
+
+func (b *metricBuilder) completeCurrentMetric() error {
+	switch mtype := b.currentMetric.MetricDescriptor.Type; mtype {
+	case metricspb.MetricDescriptor_GAUGE_DISTRIBUTION, metricspb.MetricDescriptor_CUMULATIVE_DISTRIBUTION:
+		groups := b.currentDpGroupOrdered()
+		if len(groups) == 0 {
+			// this can happen if only sum or count is added, but not following data points
+			return errors.New("no data point added to summary")
```
Comment (dinooliva) [inline receiver/prometheusreceiver/internal/metricsbuilder.go]: For Python, the Prometheus client code is incomplete for summaries and only exports sum and count - consider changing this to a log message and returning nil otherwise the entire scrape will be lost for any Python application with a Summary metric.

Comment (fivesheep) [reply receiver/prometheusreceiver/internal/metricsbuilder.go]: Good catch, I have observed similar behavior very recently from a MicroMeter' prometheus endpoint that a summary only contains sum and count. 
Returning nil, as suggested, is one way to solve the issue, however, the down side is that we might loss some important metrics which customers are interested in.  Or we can convert these kind of metrics differently into something like  two counters, or dose OpenCensus format allow to have Summary Snapshots with no Percentile values?

Comment (dinooliva) [reply receiver/prometheusreceiver/internal/metricsbuilder.go]: Additionally, the Prometheus Go client now exports empty summaries by default.

I believe that exporting summary snapshot with empty percentile values makes the most sense.

@songy23 @rghetia - wdyt?

File: receiver/prometheusreceiver/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (dinooliva) [reply receiver/prometheusreceiver/README.md]: Note that the same logic applies to histograms (count, sum, and buckets).

For summaries, it's a bit more complicated - count & sum can be handled similar to counters but percentiles are treated similar to gauges (since they capture an instant).

File: receiver/prometheusreceiver/internal/transaction_test.go [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+	t.Run("Commit Without Adding", func(t *testing.T) {
+		mcon := newMockConsumer()
+		tr := newTransaction(context.Background(), ms, mcon, testLogger)
+		if got := tr.Commit(); got != nil {
+			t.Errorf("expecting nil from Commit() but got err %v", got)
+		}
+	})
+
+	t.Run("Rollback dose nothing", func(t *testing.T) {
```
Comment (dinooliva) [inline receiver/prometheusreceiver/internal/transaction_test.go]: nit: does

File: receiver/prometheusreceiver/internal/ocastore_test.go [hunk 8]
```diff
@@ -0,0 +1,312 @@
...
+
+scrape_configs:
+- job_name: 'test'
+  scrape_interval: 80ms
+  static_configs:
+  - targets: ['%s']
+`
+
+// https://github.com/prometheus/docs/blob/master/content/docs/instrumenting/exposition_formats.md#text-format-example
+var testData1 = `
```
Comment (dinooliva) [inline receiver/prometheusreceiver/internal/ocastore_test.go]: Consider splitting the different metric types into different tests and add any missing metric types (e.g. gauge).

Also consider adding more edge cases (e.g. histogram with 1 bucket).

File: receiver/prometheusreceiver/internal/metricsbuilder.go (hunk 6 above)
Comment (rghetia) [reply receiver/prometheusreceiver/internal/metricsbuilder.go]: I agree with empty percentile values.

Comment (fivesheep) [reply receiver/prometheusreceiver/internal/metricsbuilder.go]: will do.   I am also going to drop gaugehistogram support from this pr until an official spec of this data type is provided.  as hint by @dinooliva , I have checked some code from the python prometheus client, and found it had some code related to this `gaugehistogram` type from this PR https://github.com/prometheus/client_python/pull/306 , and it seems like its a bit different from regular histograms, instead of using _count/_sum it uses  _gcount/_gsum as postfixes. the PR has also attached a draft spec link to openmetrics, the next of prometheus, which can be found from https://docs.google.com/document/d/1KwV0mAXwwbvvifBvDKH_LU1YjyXE_wxCkHNoCGq1GX0/edit#heading=h.1cvzqd4ksd23 

it also has some interesting new features like _created to indicate the start time of metrics, however, this spec is still in draft stage, and I am not sure when it will be published.

**Comment by dinooliva** [comment]
We'd like to use this code in a project that we're currently working on so I've created two issues (https://github.com/census-instrumentation/opencensus-service/issues/588 and https://github.com/census-instrumentation/opencensus-service/issues/589) for tracking the identified issues with the code.

We plan to merge the code and will fix the issues in subsequent pr's.
